tensorflow/core/kernels/bias_op.cc

```diff
@@ -268,43 +269,184 @@ class BiasGradOp : public OpKernel {
+          // So it is necessary to align the calculation by inner most dim.
+          // It is unnecessary to align by channel here, instead we can
+          // simply maintain channel id to know which channel we are
+          // dealing with.
+          const int64 stride = height * width * depth;
+          const int64 align_begin =
+              ((block_begin + stride - 1) / stride) * stride;
+          const int64 align_end = (block_end / stride) * stride;
+          int64 channel_id = block_begin / stride % channel;
+          // Dealing with front residual.
+          block_result_ptr[channel_id] +=
+              VecSumReduce(&input_ptr[block_begin], align_begin - block_begin);
+          // Init channel_id to avoid the error when align_begin == block_begin.
+          channel_id = align_begin / stride % channel;
+          // Apply the reduction
+          for (int64 i = align_begin; i < align_end; i += stride) {
+            if (channel_id < channel) {
+              // When channel_id is in channel,
+              // just add the sum of inside dim to block buffer.
+              block_result_ptr[channel_id] +=
+                  VecSumReduce(&input_ptr[i], stride);
+              channel_id++;
+            } else {
+              // When channel_id exceed the range of channel,
+              // go back to the beginning of block buffer.
+              channel_id = channel_id - channel;
+              block_result_ptr[channel_id] +=
+                  VecSumReduce(&input_ptr[i], stride);
+              channel_id++;
+            }
+          }
+          // Dealing with back residual.
+          block_result_ptr[channel_id] +=
+              VecSumReduce(&input_ptr[align_end], block_end - align_end);
+        }
+      };
+      // Run multi-threads
+      // We use Sharder::Do here to make sure each block matches one thread
+      const int total_units = num_blocks;
+      // As we have pretreated workload,
+      // set cost_per_unit to 10000 to override the defalt.
+      const int cost_per_unit = 10000;
+      Sharder::Do(
+          total_units, cost_per_unit, BiasGradWorker,
+          [&workers](Sharder::Closure c) -> void { workers->Schedule(c); },
+          max_parallelism);
+
+      //******************************************************************************
+      //  Sum block results up
+      //******************************************************************************
+      for (int64 i = 0; i < num_blocks; i++) {
+        InplaceVecAdd(result_ptr, &block_results_buffer_ptr[i * channel],
+                      channel);
       }
+      // Modified for performance tune :: end here
     }
   }
 
  private:
   TensorFormat data_format_;
+
+  // Modified for performance tune :: new funcs
+  // Apply X[0:length-1] = X[0:length-1] + Y[0:length-1];
+  inline void InplaceVecAdd(T* X, const T* Y, const int64 length) {
+    //#pragma simd
+    for (int64 i = 0; i < length; i++) {
+      X[i] = X[i] + Y[i];
```

Comment:
Do you know if this pragma works with clang? This loop can be expressed as simple Eigen expression, something like:

```
    using X = Eigen::Map<Eigen::Array<T, Dynamic, 1>>;
    using Y = Eigen::Map<const Eigen::Array<T, Dynamic, 1>>
    X(x, length) += Y(y, length);
```

I probably messed up with some constructor parameters, but similar idea is used in https://bitbucket.org/eigen/eigen/src/9c300336de9a096c7c4c3b230a6ae9f49fa17aea/unsupported/Eigen/CXX11/src/Tensor/TensorBlock.h#lines-494:506

This will be vectorized/unrolled

Reply:
Ok, repalced them with eigen apis. And I also figured there is no need to keep the `InplaceVecAdd()` and `VecSumReduce()` fuctions. I inlined them into the lambda function.